File: data/backtracking/combination-sum.cpp
#include <vector>

using namespace std;

class SolutionCombinationSum {
    vector<vector<int>> res;
    void dfs(vector<int>& candidates, int target, int i, vector<int>& cur, int total) {
        if (total == target) {
            res.push_back(cur);
            return;
        }
        if (i >= candidates.size() || total > target) return;

        cur.push_back(candidates[i]);
        dfs(candidates, target, i, cur, total + candidates[i]);
        cur.pop_back();
        dfs(candidates, target, i + 1, cur, total);
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> cur;
        dfs(candidates, target, 0, cur, 0);
        return res;
    }
};

File: data/combination/combination-sum.py
def combinationSum(candidates: list[int], target: int) -> list[list[int]]:
    """
    Given an array of distinct integers candidates and a target integer target, 
    return a list of all unique combinations where the chosen numbers sum to target.
    """
    res = []
    def dfs(i, cur, total):
        if total == target:
            res.append(cur[:])
            return
        if i >= len(candidates) or total > target:
            return
        cur.append(candidates[i])
        dfs(i, cur, total + candidates[i])
        cur.pop()
        dfs(i + 1, cur, total)

    dfs(0, [], 0)
    return res
